trio/_core/_asyncgen.py

```diff
@@ -0,0 +1,323 @@
+
+    """
+
+    owner = AsyncGenOwner(owner_name)
+    prev_hooks = sys.get_asyncgen_hooks()
+
+    @wraps(firstiter)
+    def firstiter_if_ours(agen):
+        if is_ours(agen):
+            agen.ag_frame.f_locals["@trio_asyncgen_owner"] = owner
+            firstiter(agen)
+        elif prev_hooks.firstiter is not None:
+            prev_hooks.firstiter(agen)
+
+    @wraps(finalizer)
+    def finalizer_if_ours(agen):
+        if agen.ag_frame.f_locals.get("@trio_asyncgen_owner") is owner:
+            finalizer(agen)
+        elif prev_hooks.finalizer is not None:
+            prev_hooks.finalizer(agen)
+        else:
+            fallback_finalizer_when_no_asyncgen_hooks_have_claimed(agen)
+
+    sys.set_asyncgen_hooks(firstiter=firstiter_if_ours, finalizer=finalizer_if_ours)
+    try:
+        yield
+    finally:
+        sys.set_asyncgen_hooks(*prev_hooks)
+
+
+@attr.s(eq=False)
+class AsyncGeneratorManager:
+    """Manage async generator finalization for a single Trio run.
+
+    Surround the entire run in 'with manager.hooks_context():';
+    within this scope the async generator hooks will be active. Then,
+    call 'manager.spawn_reaper()' once the system nursery is available;
+    this will run reaper() as a system task to host the actual
+    finalizations.
+    """
+
+    # Entry queue for reentering the run loop from our finalizer hook
+    entry_queue = attr.ib()
+
+    # WeakSet of still-referenced async generators. We will finalize all of these
+    # after the main task exits.
+    asyncgens = attr.ib(init=False, factory=weakref.WeakSet)
+
+    # Nursery created by the reap task, into which we will spawn instances
+    # of self.finalize.
+    nursery = attr.ib(init=False, default=None)
+
+    # If an async generator needs to be finalized before the reap task has
+    # started, we remember it in this list. Contains tuples of arguments for
+    # self.start_finalize.
+    pending_finalize_calls = attr.ib(init=False, factory=list)
+
+    # This list contains one None value per in-flight run_sync_soon call.
+    # (Appending and popping a list are signal-safe, while mutating an integer
+    # is not.)
+    inflight_starts = attr.ib(init=False, factory=list)
+
+    # The reaper task waiting for inflight_starts to be empty, or None if no
+    # one is waiting.
+    task_awaiting_quiescence = attr.ib(init=False, default=None)
+
+    # True once we have kicked off finalization of all remaining asynvgens
+    # (at main task exit) and can't reasonably handle more.
+    closing = attr.ib(init=False, default=False)
+
+    def firstiter_hook(self, agen):
+        if self.closing:
+            warnings.warn(
+                f"New async generator {name_asyncgen(agen)!r} was created after "
+                f"Trio had already finalized outstanding async generators, so we "
+                f"will not be able to manage its lifetime.",
+                RuntimeWarning,
+                stacklevel=2,
+            )
```

Comment:
Yep, they are -- I went with a warning because it would technically be a backcompat break to make this an error, but it almost certainly wouldn't matter in practice, and I do prefer the error on aesthetic grounds. (Also, one can catch the exception and call `__anext__` again, since at that point it's not the first iteration anymore and doesn't call the hook.)